keps/sig-scheduling/20190926-default-even-pod-spreading.md

```diff
@@ -0,0 +1,278 @@
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+With `EvenPodsSpreading`, workload authors can define spreading rules for their loads based on the
+topology of the clusters. 
+
+By introducing configurable default spreading rules, workloads can be spread in the topology of
+the cluster according to opinionated rules set by a cluster operator, without having to explicitly
+define them. And workloads with specific needs can still override the rules by defining them in
+their PodSpec.
+
+## Motivation
+
+In order for a workload (pod) to use `EvenPodsSpreading`:
+
+1. Authors have to have an idea of the underlying topology.
+1. PodSpecs become less portable if their spreading rules are tailored to a specific topology.
+
+On the other hand, cluster operators know the underlying topology of the cluster, which makes
+them suitable to provide default spreading rules for all workloads in their cluster.
+
+### Goals
+
+- Cluster operators can define default spreading rules for pods that don't provide any.
+- Workloads are spread with the default rules if they belong to the same service, controller,
+replica set or stateful set,
+and if they don't define `TopologySpreadConstraints`.
+
+### Non-Goals
+
+- Removal of `SelectorSpreadPriority`, `ServiceSpreadingPriority` or `ServiceAntiAffinity` priorities.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, I want to set default spreading rules for workloads in the cluster.
+Currently, `SelectorSpreadPriority` provides a canned priority that spreads across nodes
+and zones (`failure-domain.beta.kubernetes.io/zone`). However, the nodes in my cluster have
+custom topology keys (for physical host, rack, etc.).
+
+#### Story 2
+
+As a workload author, I want to spread the workload in the cluster, but:
+(1) I don't know the topology of the cluster I'm running on.
+(2) I want to be able to run my PodSpec in different clusters (on-prem and cloud).
+
+### Implementation Details/Notes/Constraints
+
+Note that a priority given by default `EvenPodsSpreading` rules could conflict with
+`SelectorSpreadingPriority`.
+Operators can disable `SelectorSpreadingPriority`. But once default rules for `EvenPodsSpreading` is GA,
+we can consider removing `SelectorSpreadingPriority` and replacing it by an equivalent
+k8s default to the default rules for `EvenPodsSpreading`.
+
+### Risks and Mitigations
+
+`EvenPodsSpreading` has some overhead and we currently ensure that pods that don't use the
+feature get minimally affected. After default rules for `EvenPodsSpreading` is rolled out,
```

Comment:
Ditto. What does "default rules" mean? Are you trying to mean **this feature** with partial APIs of `EvenPodsSpreading`? If it's the case, we should explicitly reword to the API "topologySpreadConstraint".